### ISSUE TITLE
Notification scheduling on Android 6 and Android 7 fails 

### ISSUE THREAD

**Issue by tigertore** [MaikuB/flutter_local_notifications#1149]
First of all, thanks for your great work on this plugin.
The plugin works fine on Android 8 and above, but we see some crashes and un-handled exceptions in crashlytics on Android 6 and Android 7.
I have not been able to reproduce the errors myself, but it seems to be related to the initialization of the AndroidThreeTen library.

Plugin version : flutter_local_notifications: 5.0.0+1

Here is a summary of the errors I see in crashlytics:

#########################################################################

Unhandled and Non-fatal exceptions :

Android 7.1.1, Xperia Z5 Compact, 26 Apr 2021, 20:39:20
Android 7.0, Galaxy S6 edge, 26 Apr 2021, 20:06:56
Android 7.0, Xperia L1, 26 Apr 2021, 17:21:57
Android 6.0, Honor 7, 26 Apr 2021, 13:46:18

Non-fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: PlatformException(error, No time-zone data files registered, null, n.b.a.y.g: No time-zone data files registered
	at n.b.a.y.i.a(SourceFile:3)
	at n.b.a.y.i.b(SourceFile:3)
	at n.b.a.s.a(SourceFile:3)
	at n.b.a.q.a(SourceFile:10)
	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.zonedScheduleNotification(SourceFile:7)
	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.zonedSchedule(SourceFile:5)
	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.onMethodCall(SourceFile:13)
	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(SourceFile:2)
	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(SourceFile:5)
	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(SourceFile:2)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(MessageQueue.java:331)
	at android.os.Looper.loop(Looper.java:219)
	at android.app.ActivityThread.main(ActivityThread.java:6274)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
). Error thrown null.
       at StandardMethodCodec.decodeEnvelope(StandardMethodCodec.java:581)
       at MethodChannel._invokeMethod(MethodChannel.java:158)
       at AndroidFlutterLocalNotificationsPlugin.zonedSchedule(AndroidFlutterLocalNotificationsPlugin.java:137)
       at FlutterLocalNotificationsPlugin.zonedSchedule(FlutterLocalNotificationsPlugin.java:301)
	   
	   

Android 7.1.1, Xperia Z5 Compact, 26 Apr 2021, 20:34:31
Android 7.0, Galaxy S6 edge, 26 Apr 2021, 20:06:56
Android 7.0, Xperia L1, 26 Apr 2021, 17:21:56
Android 6.0, P8 Lite, 26 Apr 2021, 13:43:58

Non-fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: PlatformException(error, Already initialized, null, java.lang.IllegalStateException: Already initialized
	at n.b.a.y.h.a(SourceFile:4)
	at e.e.a.a.a(SourceFile:3)
	at e.e.a.a.a(SourceFile:1)
	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.initAndroidThreeTen(SourceFile:2)
	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.initialize(SourceFile:4)
	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.onMethodCall(SourceFile:17)
	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(SourceFile:2)
	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(SourceFile:5)
	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(SourceFile:2)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(MessageQueue.java:331)
	at android.os.Looper.loop(Looper.java:219)
	at android.app.ActivityThread.main(ActivityThread.java:6274)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
). Error thrown null.
       at StandardMethodCodec.decodeEnvelope(StandardMethodCodec.java:581)
       at MethodChannel._invokeMethod(MethodChannel.java:158)
       at AndroidFlutterLocalNotificationsPlugin.initialize(AndroidFlutterLocalNotificationsPlugin.java:85)
	   
	   
	   
Android 7.0, honor 6A Pro, 26 Apr 2021, 15:58:17
Android 6.0, P8 Lite, 26 Apr 2021, 13:43:58

Non-fatal Exception: io.flutter.plugins.firebase.crashlytics.FlutterError: PlatformException(error, No time-zone data files registered, null, n.b.a.y.g: No time-zone data files registered
	at n.b.a.y.i.a(SourceFile:3)
	at n.b.a.y.i.b(SourceFile:3)
	at n.b.a.s.a(SourceFile:3)
	at n.b.a.q.a(SourceFile:10)
	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.getNextFireDateMatchingDateTimeComponents(SourceFile:14)
	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.zonedSchedule(SourceFile:4)
	at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.onMethodCall(SourceFile:13)
	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(SourceFile:2)
	at io.flutter.embedding.engine.dart.DartMessenger.handleMessageFromDart(SourceFile:5)
	at io.flutter.embedding.engine.FlutterJNI.handlePlatformMessage(SourceFile:2)
	at android.os.MessageQueue.nativePollOnce(Native Method)
	at android.os.MessageQueue.next(MessageQueue.java:356)
	at android.os.Looper.loop(Looper.java:138)
	at android.app.ActivityThread.main(ActivityThread.java:6623)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:942)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:832)
). Error thrown null.
       at StandardMethodCodec.decodeEnvelope(StandardMethodCodec.java:581)
       at MethodChannel._invokeMethod(MethodChannel.java:158)
       at AndroidFlutterLocalNotificationsPlugin.zonedSchedule(AndroidFlutterLocalNotificationsPlugin.java:137)
       at FlutterLocalNotificationsPlugin.zonedSchedule(FlutterLocalNotificationsPlugin.java:301)
	   

#########################################################################

Crashes :

App seems to be in background.

Android 7.0, Galaxy S6, 26 Apr 2021, 14:50:00
Android 6.0.1, Xperia Z2, 26 Apr 2021, 14:48:00
Android 6.0.1, Galaxy S5, 26 Apr 2021, 10:11:00

Fatal Exception: java.lang.RuntimeException: Unable to start receiver com.dexterous.flutterlocalnotifications.ScheduledNotificationReceiver: n.b.a.y.g: No time-zone data files registered
       at android.app.ActivityThread.handleReceiver(ActivityThread.java:3316)
       at android.app.ActivityThread.-wrap20(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1715)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6682)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)
	   
	   Caused by n.b.a.y.g: No time-zone data files registered
       at org.threeten.bp.zone.ZoneRulesProvider.getProvider(ZoneRulesProvider.java:165)
       at org.threeten.bp.zone.ZoneRulesProvider.getRules(ZoneRulesProvider.java:122)
       at org.threeten.bp.ZoneRegion.ofId(ZoneRegion.java:143)
       at org.threeten.bp.ZoneId.of(ZoneId.java:358)
       at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.getNextFireDateMatchingDateTimeComponents(FlutterLocalNotificationsPlugin.java:861)
       at com.dexterous.flutterlocalnotifications.FlutterLocalNotificationsPlugin.zonedScheduleNextNotificationMatchingDateComponents(FlutterLocalNotificationsPlugin.java:813)
       at com.dexterous.flutterlocalnotifications.ScheduledNotificationReceiver.onReceive(ScheduledNotificationReceiver.java:50)
       at android.app.ActivityThread.handleReceiver(ActivityThread.java:3309)
       at android.app.ActivityThread.-wrap20(ActivityThread.java)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1715)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:154)
       at android.app.ActivityThread.main(ActivityThread.java:6682)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)

**Comment by MaikuB** [comment]
Unless ways to reproduce this can be provided, I won't be able to track down the issue to fix it. I've already tried things like trying to initialise the library twice and don't see the error mentioned where the AndroidThreeTen library reports that it's already been initialised.

The only thing I can do that would solve this in a future version of the plugin is to have it use Android Gradle Plugin 4.0+ so that the dependency on the AndroidThreeTen library can be removed. This would require apps to update as well though I'm not sure if issues like [this](https://github.com/flutter/flutter/issues/58247) still exist

**Comment by tigertore** [comment]
I have analysed more logs from crashlytics and we probably have a race-condition in our code.
From the logs it seems that the app sometimes calls zonedSchedule() before initialize().
I have identified a flow, in our code, that can trigger this situation. 
When I mock the data and get this flow, I get the excact same exceptions as found in crashlytics,
and they only appear on Android 6 and 7.

The bad thing is that the initialize() fails with "Already initialized", 
so all calls to zonedSchedule() will fail with "No time-zone data files registered"

If I then put the app into the background and the ScheduledNotificationReceiver is triggered,
we will probably get the crash seen above. I have not tested this.
